### ISSUE TITLE
Better support unit-testing of reducers that handle lifecycle actions

### ISSUE THREAD

**Issue by lukasgebhard** [reduxjs/redux-toolkit#872]
## Context
Using `createAsyncThunk`, I created a thunk `fetchData`to fetch data from the network. I unit-test the (non-trivial) thunk logic and the lifecycle reducers seperately, that is, I have unit tests to 

- Check if the thunk dispatches the right actions, and
- Check if the reducers update the store correctly for each of the three lifecycle actions.

## Problem
To test the reducer logic for any of the lifecycle actions, say the `fulfilled` action, I need to dispatch the action. I see three options to create the action, none of which seems ideal:

1) One option is to call `fetchData.fulfilled`. However, the action creator's parameters are not explained anywhere in the official docs, and the number of parameters was silently changed in release 1.5.0. This suggests that the action creator is not intended to be called from outside of RTK.
1) Instead, I could easily write my own action creator of course. However, I guess it would be better if RTK provided it as it would likely be beneficial for many RTK users. After all, unit-testing reducers is quite a common use case :-)
1) A third option is to mock the data fetching to have the thunk dispatch the lifecycle actions. This is not optimal either because unit-testing reducers should be independent of thunks.

## TLDR
I think it would be best if the three lifecycle action creators exposed by a thunk would be an official part of RTK's API.

**Comment by lukasgebhard** [comment]
For the meantime, I chose option (2), that is, I wrote action creators to mock the lifecycle actions:

```typescript
export function mockPendingAction<ThunkArg>(
  typePrefix: string,
  thunkArg: ThunkArg
): PendingAction<ThunkArg> {
  return {
    type: typePrefix + '/pending',
    payload: undefined,
    meta: {
      requestId: '0',
      arg: thunkArg
    }
  }
}
export function mockFulfilledAction<PromiseResult, ThunkArg>(
  typePrefix: string,
  promiseResult: PromiseResult,
  thunkArg: ThunkArg
): FulfilledAction<ThunkArg, PromiseResult> {
  return {
    type: typePrefix + '/fulfilled',
    payload: promiseResult,
    meta: {
      requestId: '0',
      arg: thunkArg
    }
  }
}
export function mockRejectedAction<ThunkArg>(
  typePrefix: string,
  error: SerializedError,
  thunkArg: ThunkArg
): RejectedAction<ThunkArg> {
  return {
    type: typePrefix + '/rejected',
    payload: undefined,
    error: error,
    meta: {
      requestId: '0',
      arg: thunkArg,
      aborted: false,
      condition: false
    }
  }
}
export function mockRejectedWithValueAction<ThunkArg, RejectedValue>(
  typePrefix: string,
  rejectedValue: RejectedValue,
  thunkArg: ThunkArg
): RejectedWithValueAction<ThunkArg, RejectedValue> {
  return {
    type: typePrefix + '/rejected',
    payload: rejectedValue,
    error: { message: 'Rejected' },
    meta: {
      requestId: '0',
      arg: thunkArg,
      aborted: false
    }
  }
}
```

However, as redux-toolkit does not export the action types ([`PendingAction`, `FulfilledAction`, `RejectedAction`, `RejectedWithValueAction`](https://redux-toolkit.js.org/api/createAsyncThunk#promise-lifecycle-actions)), I need to re-define them. I therefore recommend exporting the action types unless you decide to expose the lifecycle action creators (which still seems to be the best option).

**Comment by markerikson** [comment]
I guess I'm not really sure what the request is here. What specific changes are you asking for?

**Comment by markerikson** [comment]
Hmm. Still not quite sure what you mean by that.

The relevant action creators are exposed by attaching them to the thunk.  What about that is not sufficient?

**Comment by phryneas** [comment]
@markerikson they have actually never been part of the "official API" so they are not perfectly usable - `reject` can at the moment not be used without `RejectWithValue` which we do not export.
I'm personally not a fan of this type of testing and would suggest that people test the resulting thunk *as one entity*, but there have been enough requests by now. Maybe we should change the signature of `reject` and make those action creators part of the official api.

**Comment by markerikson** [comment]
So what concrete changes are actually needed, then?

**Comment by phryneas** [comment]
@markerikson undoing this signature change:
https://github.com/reduxjs/redux-toolkit/commit/6e95fb0d7bfcca4690ec4b35a6dccaf9cfa2dc59#diff-72c9100d5dcb08a309727a42a354cd576dcf318c3d0820c98ad26ef2bf6392e6L324-L332

**Comment by markerikson** [comment]
Okay, clearly there's something about this discussion I'm not grasping.

Why does the `rejected` signature need to be reverted to what it was previously?  How does that relate to this discussion?

**Comment by phryneas** [comment]
Before, I could be called `(error, requestId, arg)` or `(null, requestId, arg, rejectedPayload)`.

Now it would be called
`(error, requestId, arg)` or `(RejectedClass(rejectedPayload), requestId, arg)`.

But `RejectedClass` is not exported, so the "rejected with value" case can not be reproduced in unit tests (other than just creating that action by hand).

**Comment by markerikson** [comment]
So what's the right solution here? Export `RejectedClass`? Export `rejectWithValue?` Do something else with the action creator signature?

**Comment by phryneas** [comment]
Just undoing the signature change would probably be enough. Before 1.5 I had not considered it as a public interface, because it was never documented, but the signature before apparently was used and worked fine for the purpose.

**Comment by markerikson** [comment]
Resolved in #1083 .
